WalletWasabi.Fluent.Generators/UIContextGenerator.cs

```diff
@@ -0,0 +1,313 @@
+
+		var ctors =
+			classDeclaration.ChildNodes()
+							.OfType<ConstructorDeclarationSyntax>()
+							.ToList();
+
+		foreach (var ctor in ctors)
+		{
+			if (!classDeclaration.GetUiContextReferences(semanticModel).Any())
+			{
+				if (!classDeclaration.IsAbstractClass(semanticModel) && ctor.IsPublic())
+				{
+					yield return ctor;
+				}
+			}
+			else
+			{
+				var ctorArgs =
+					  ctor.ParameterList.Parameters
+						  .Select(x => x.Identifier.ValueText)
+						  .ToList();
+
+				var hasCtorArgs = ctorArgs.Any();
+				var ctorArgsString = string.Join(",", ctorArgs);
+				var ctorString =
+					hasCtorArgs
+					? $": this({ctorArgsString})"
+					: "";
+
+				var parameterUsings =
+					ctor.ParameterList.Parameters
+									  .Where(p => p.Type is not null)
+									  .Select(p => semanticModel.GetTypeInfo(p.Type))
+									  .Where(t => t.Type is not null)
+									  .Select(t => $"using {t.Type.ContainingNamespace.ToDisplayString()};")
+									  .ToList();
+
+				var parametersString = "UiContext uiContext";
+
+				var uiContextParameter =
+					SyntaxFactory.Parameter(SyntaxFactory.Identifier("uiContext")
+														 .WithLeadingTrivia(SyntaxFactory.Space))
+								 .WithType(SyntaxFactory.ParseTypeName("UiContext"));
+
+				if (hasCtorArgs)
+				{
+					parametersString += ", ";
+				}
+
+				parametersString =
+					ctor.ParameterList.Parameters.Insert(0, uiContextParameter).ToFullString();
+
+				var usings = string.Join(Environment.NewLine, parameterUsings.Distinct().OrderBy(x => x));
+
+				var code =
+	$$"""
+{{usings}}
+using WalletWasabi.Fluent.Models.UI;
+
+namespace {{namespaceName}};
+
+partial class {{className}}
+{
+    public {{className}}({{parametersString}}){{ctorString}}
+    {
+	    UiContext = uiContext;
+    }
+}
+""";
+
+				var sourceText = SourceText.From(code, Encoding.UTF8);
+				context.AddSource(fileName, sourceText);
+
+				var tree = SyntaxFactory.ParseSyntaxTree(sourceText);
+
+				var newConstructor =
+					tree.GetRoot()
+						.DescendantNodes()
+						.OfType<ConstructorDeclarationSyntax>()
+						.FirstOrDefault();
+
+				yield return newConstructor;
+			}
+		}
+	}
+
+	private static void GenerateFluentNavigation(SourceProductionContext context, Compilation compilation, IEnumerable<ConstructorDeclarationSyntax> ctors)
+	{
+		var namespaces = new List<string>();
+		var methods = new List<string>();
+
+		var newSyntaxTrees =
+			ctors.Select(x => x.SyntaxTree)
+				 .Where(x => !compilation.ContainsSyntaxTree(x))
+				 .ToList();
+
+		compilation = compilation.AddSyntaxTrees(newSyntaxTrees);
+
+		foreach (var ctor in ctors)
+		{
+			var semanticModel = compilation.GetSemanticModel(ctor.SyntaxTree);
+
+			if (ctor.Parent is not ClassDeclarationSyntax cls)
+			{
+				continue;
+			}
+
+			if (!cls.IsRoutableViewModel(semanticModel))
+			{
+				continue;
+			}
+
+			var viewModelTypeInfo =
+				semanticModel.GetDeclaredSymbol(ctor.Parent as ClassDeclarationSyntax);
```

Comment:
Are we sure Parent is always a `ClassDeclarationSyntax`?

Reply:
@soosr that is checked in line 191:

```csharp
if (ctor.Parent is not ClassDeclarationSyntax cls)
			{
				continue;
			}
```

Anyways, I changed that code to: `semanticModel.GetDeclaredSymbol(cls);` in order to avoid additional cast.